src/App/Pages/Accounts/TwoFactorPageViewModel.cs

```diff
@@ -326,33 +326,36 @@ public async Task SubmitAsync(bool showLoading = true)
 
                 if (decryptOptions?.TrustedDeviceOption != null)
                 {
-                    // If user doesn't have a MP, but has reset password permission, they must set a MP
-                    if (!decryptOptions.HasMasterPassword &&
-                        decryptOptions.TrustedDeviceOption.HasManageResetPasswordPermission)
-                    {
-                        StartSetPasswordAction?.Invoke();
-                    }
-                    else if (result.ForcePasswordReset)
-                    {
-                        UpdateTempPasswordAction?.Invoke();
-                    }
-                    else if (await _deviceTrustCryptoService.IsDeviceTrustedAsync())
+                    if (await _deviceTrustCryptoService.IsDeviceTrustedAsync())
                     {
+                        // If we have a device key but no keys on server, we need to remove the device key
                         if (decryptOptions.TrustedDeviceOption.EncryptedPrivateKey == null && decryptOptions.TrustedDeviceOption.EncryptedUserKey == null)
                         {
                             await _deviceTrustCryptoService.RemoveTrustedDeviceAsync();
                             StartDeviceApprovalOptionsAction?.Invoke();
+                            return;
                         }
-                        else
+                        // If user doesn't have a MP, but has reset password permission, they must set a MP
+                        if (!decryptOptions.HasMasterPassword &&
+                            decryptOptions.TrustedDeviceOption.HasManageResetPasswordPermission)
                         {
-                            _syncService.FullSyncAsync(true).FireAndForget();
-                            await TwoFactorAuthSuccessAsync();
+                            StartSetPasswordAction?.Invoke();
+                            return;
                         }
+                        // Update temp password only if the device is trusted and therefore has a decrypted User Key set
+                        if (result.ForcePasswordReset)
+                        {
+                            UpdateTempPasswordAction?.Invoke();
+                            return;
+                        }
+
+                        // Device is trusted and has keys, so we can decrypt
+                        _syncService.FullSyncAsync(true).FireAndForget();
+                        await TwoFactorAuthSuccessAsync();
+                        return;
                     }
-                    else
-                    {
-                        StartDeviceApprovalOptionsAction?.Invoke();
-                    }
+
+                    StartDeviceApprovalOptionsAction?.Invoke();
                     return;
```

Comment:
♻️ Most of this logic is exactly the same as in `LoginSSOViewModel`, could we refactor it so we reuse the code and when changes are needed (like in this case) we don't have to modify both places? If this is time sensitive, we could create a tech-debt task (can't remember if one exists already)

Reply:
Thought about it as well, but i think we are on a time crunch better to create a ticket for that

Reply:
Created PM-3502